src/draw/sw/lv_draw_sw_blur.c

```diff
@@ -198,7 +198,7 @@ void lv_draw_sw_blur(lv_draw_task_t * t, const lv_draw_blur_dsc_t * dsc, const l
             }
 
             buf_line = lv_draw_buf_goto_xy(t->target_layer->draw_buf, x_end, y);
-            blur_3_bytes_init(sum, buf_line, sample_len_limited, - px_size * skip_cnt);
+            blur_3_bytes_init(sum, buf_line, sample_len_limited, -(int32_t)px_size * skip_cnt);
```

Comment:
What about this? Avoids the cast

```suggestion
            blur_3_bytes_init(sum, buf_line, sample_len_limited, px_size * -skip_cnt);
```

Reply:
@AndreCostaaa 

cc:  @kisvegabor 

> What about this? Avoids the cast

Fair question.  This is one of those areas of C compiler frailty and difficult-to-see bugs.  If you want to look it up, it's under the topic of [C Compiler Implicit Type Promotions](https://en.cppreference.com/w/c/language/conversion.html).  It doesn't work because according to C specification, the compiler then silently converts the SIGNED SIDE of the multiplication to `uint32_t` before doing the multiplication, which then dramatically alters the value back to a very large positive number when it started out being negative!

Specifically, if px_size == 4 and skip_cnt == 128, the expression `px_size * -skip_cnt`:
1. Negates signed 128 to -128.
2. Silently "promotes" the type of the -128 value to a `uint32_t` resulting in the intermediate value `4294967168` (0xFFFFFE00) on the right side of the multiplication.
3. Then the multiplication happens multiplying that X 4U which then OVERFLOWS the `uint32_t`.
4. Then converting the least-significant 32 bits of the result (that binary pattern) BACK to signed 32-bit value again results in -512.

It's "by accident" the same result, but is VERY unintentional method of getting there, and is probably not portable.

The correct method is to SUPPRESS the silent type promotion of the right side of the `*` operator by converting the `uint32_t px_size` to a signed value matching the type of the `int32_t skip_cnt` variable.  Then both values are signed when the multiplication happens.  `-(int32_t)px_size * skip_cnt` results in:
1. 4U is converted to 4 (signed).
2. Value is negated to -4.
3. Signed multiplication:  -4 * 128 (both signed) = -512.

That silent type promotion is suppressed by casting the `uint32_t` to the same signed type as `skip_cnt`.  I thought about:
```c
blur_3_bytes_init(sum, buf_line, sample_len_limited, -((int32_t)px_size * skip_cnt));
```
which delays the negation step until the end, but to me it just seemed to make it less readable while not changing the outcome.  Either way, the result is the same, and so I opted for the more readable version.

Note that if a coding standard favoring type prefixes is used, how the bug "jumps out at you".
```c
blur_3_bytes_init(sum, buf_line, sample_len_limited, -ui32_px_size * i32_skip_cnt);
// the above bug (negating an unsigned value) is plainly visible, with the solution being
blur_3_bytes_init(sum, buf_line, sample_len_limited, -(int32_t)ui32_px_size * i32_skip_cnt);
```

You can see the intermediate value computed in this example:
```c
#include <stdio.h>
#include <stdint.h>
int main()
{
    uint32_t px_size = 4;
    int32_t skip_cnt = 128;
    int32_t result1;
    int32_t result2;
    uint32_t result2_intermediate;
    result1 = -(int32_t)px_size * skip_cnt;
    result2_intermediate = px_size * -skip_cnt;
    result2 = px_size * -skip_cnt;
    printf("Result1: %d\n", result1);
    printf("Intermediate: %u\n", result2_intermediate);
    printf("Result2: %d\n", result2);
}
```
```
Result1: -512
Intermediate: 4294966784
Result2: -512
```

Reply:
> It's "by accident" the same result, but is VERY unintentional method of getting there, and is probably not portable.

Not by accident, it's by design and it will be very portable as it's defined in the C standard

If it weren't portable every single C code would have this bug

If you interpret your "very big" unsigned value as signed value it will look correct. Your computer doesn't care about how you're interpreting the data, it just shifts 0s and 1s around

Reply:
> Not by accident, it's by design...

cc:  @kisvegabor 

The trouble—and the bug—is that when the math sequence begins, it overflows the integer, and this will always and forever be considered a bug.

Reply:
Hi, André!  ( @AndreCostaaa  )

Towards being on the same page about this:  after a couple of days, I got the feeling that perhaps you weren't understanding what I was trying to explain.  (I understood your points, but there is a reason I think we should have a closer look at this.)

## The Original Problem:

```c
blur_3_bytes_init(sum, buf_line, sample_len_limited, px_size * -skip_cnt);
```

Expression -px_size * skip_cnt rejected by MSVC as an error:  negating an unsigned value is still unsigned.  (px_size = uint32_t, skip_cnt = int32_t)

This becomes more visible if I simply prefix the variable names with their types:
```c
-ui32_px_size * i32_skip_cnt
```

However, what you may not have observed is an additional hidden problem, which you can see if you plug in these respective values in the 2 variables:  4U (unsigned) and 128 (signed):
```
    Action                    Result Value                            Expression Type
1.  "negating" the 4U         4294967292 (0xFFFFFFFC)                 uint32_t  <== first problem
2.  promote 128 to 128U       128U                                    uint32_t
3.  unsigned mult 4U * 128U   549755813376 (0x7FFFFFFE00 overflow!)   uint32_t  <== hidden problem
4.  pass to a int32_t arg     -512 (by truncation)                    int32_t
```

The correct fix suppresses step 2 above by having both values being signed:
```c
-(int32_t)ui32_px_size * i32_skip_cnt
```

Resulting in:
```
    Action                      Result Value               Expression Type
1.  demote 4U to 4 (signed)     4                          int32_t
2.  negating 4                  -4                         int32_t
3.  signed mult -4 * 128        -512                       int32_t
4.  pass to a int32_t arg       -512 (no alteration)       int32_t
```

## Your Suggestion

You suggested this instead "to save a cast":
```c
ui32_px_size * -i32_skip_cnt
```

This, however, this not only leaves the "hidden problem" intact, but introduces a new problem:
```
    Action                          Result Value                           Expression Type
1.  negate 128 (signed)             -128                                   int32_t
2.  promote -128 to unsigned        4294967168U (0xFFFFFF80)               uint32_t  <== new problem
3.  unsigned mult 4U * 4294967168U  17179868672U (0x3FFFFFE00 overflow!)   uint32_t  <== hidden problem
4.  pass to a int32_t arg           -512 (by truncation)                   int32_t
```

Does that make more sense why I objected?

**Your point:**

In the expected value ranges, the receiving function would continue to get the intended value.  Yes, I see that.  But over a long career, much of which was living and breathing bits and bytes, I have also seen code like this hide unintended problems that surface in the future because of some code change "over there", or the numeric range increases with hardware changes.

**My point:**

Doing so by overflowing an integer is what causes development teams to reject using libraries.  Plus development teams that require high-reliability code, seeing the original code (or the code suggested) "go nuts" about stuff like this.  And they should!  From their view, they require it to be "done right" or else it gives them additional reasons to write their own graphics library rather than to rely on "an outside library".  Reason:  hidden future problems hide in code like that.  It's the very reason why some teams turn on ALL WARNINGS and some even MAKE WARNINGS INTO ERRORS during compiling, because such warnings are helpful towards catching unwanted future problems.  And frankly, I think we should too.

Kind regards,
Vic